aggregators/aggregator_test.go

```diff
@@ -1075,163 +1075,438 @@ func TestHarvest(t *testing.T) {
-	batch := modelpb.Batch{
-		{
-			Event: &modelpb.Event{
-				Outcome:  "success",
-				Duration: uint64(txnDuration),
-			},
-			Transaction: &modelpb.Transaction{
-				Name:                "foo",
-				Type:                "txtype",
-				RepresentativeCount: 1,
-			},
-			Service: &modelpb.Service{Name: "svc"},
-			Labels: modelpb.Labels{
-				"department_name": &modelpb.LabelValue{Global: true, Value: "apm"},
-				"organization":    &modelpb.LabelValue{Global: true, Value: "observability"},
-				"company":         &modelpb.LabelValue{Global: true, Value: "elastic"},
-				"mylabel":         &modelpb.LabelValue{Global: false, Value: "myvalue"},
-			},
-			NumericLabels: modelpb.NumericLabels{
-				"user_id":        &modelpb.NumericLabelValue{Global: true, Value: 100},
-				"cost_center":    &modelpb.NumericLabelValue{Global: true, Value: 10},
-				"mynumericlabel": &modelpb.NumericLabelValue{Global: false, Value: 1},
-			},
-		},
+
+	runTest := func(t *testing.T, batch modelpb.Batch, expected []*modelpb.APMEvent) {
+		var events []*modelpb.APMEvent
+		agg, err := New(
+			WithDataDir(t.TempDir()),
+			WithLimits(Limits{
+				MaxSpanGroups:                         1000,
+				MaxSpanGroupsPerService:               100,
+				MaxTransactionGroups:                  100,
+				MaxTransactionGroupsPerService:        10,
+				MaxServiceTransactionGroups:           100,
+				MaxServiceTransactionGroupsPerService: 10,
+				MaxServices:                           10,
+			}),
+			WithProcessor(sliceProcessor(&events)),
+			WithAggregationIntervals([]time.Duration{time.Second}),
+		)
+		require.NoError(t, err)
+		require.NoError(t, agg.AggregateBatch(
+			context.Background(),
+			EncodeToCombinedMetricsKeyID(t, "ab01"),
+			&batch,
+		))
+		require.NoError(t, agg.Close(context.Background()))
+
+		assert.Empty(t, cmp.Diff(
+			expected,
+			events,
+			cmpopts.IgnoreTypes(netip.Addr{}),
+			cmpopts.SortSlices(func(a, b *modelpb.APMEvent) bool {
+				// Sort by Metricset.Name, then by (sorted) labels.
+				// labels are sorted by key before comparing.
+				// labels keys are compared and when equal, Value are compared.
+				// If Value is equal, compare each element in Values.
+
+				// handle base case, we can sort by Metricset.Name
+				if a.Metricset.Name != b.Metricset.Name {
+					return a.Metricset.Name < b.Metricset.Name
+				}
+
+				// otherwise sort by sorted labels
+				akeys := make([]string, 0, len(a.Labels))
+				for k := range a.Labels {
+					akeys = append(akeys, k)
+				}
+				sort.Strings(akeys)
+
+				bkeys := make([]string, 0, len(b.Labels))
+				for k := range a.Labels {
+					bkeys = append(bkeys, k)
+				}
+				sort.Strings(bkeys)
+
+				for i := 0; i < len(akeys); i++ {
+					if akeys[i] != bkeys[i] {
+						return akeys[i] < bkeys[i]
+					}
+
+					akey := akeys[i]
+					if a.Labels[akey].Value != "" && a.Labels[akey].Value != b.Labels[akey].Value {
+						return a.Labels[akey].Value < b.Labels[akey].Value
+					}
```

Comment:
For reviewers: these tests are not directly used with the current tests. I added them anyway to make this sorting function more robust and prevent unexpected failures in the future if we update the test cases.